pkg/chartsync/git.go

```diff
@@ -0,0 +1,407 @@
+
+		releaseQueue: queue,
+	}
+}
+
+// Run starts the sync of `GitChartSource`s.
+func (c *GitChartSourceSync) Run(stopCh <-chan struct{}, errCh chan error, wg *sync.WaitGroup) {
+	c.logger.Log("info", "starting sync of git chart sources")
+
+	wg.Add(1)
+	go func() {
+		defer runtime.HandleCrash()
+		defer func() {
+			c.mirrors.StopAllAndWait()
+			wg.Done()
+		}()
+
+		for {
+			select {
+			case changed := <-c.mirrors.Changes():
+				for mirrorName := range changed {
+					mirror, ok := c.mirrors.Get(mirrorName)
+
+					// Get the HelmReleases that make use of the mirror.
+					hrs, err := c.getHelmReleasesForMirror(mirrorName)
+					if err != nil {
+						c.logger.Log("error", "failed to get git chart sources for mirror", "mirror", mirrorName, "err", err)
+						continue
+					}
+
+					// We received a signal from a no longer existing
+					// mirror.
+					if !ok {
+						if len(hrs) == 0 {
+							// If there are no references to it either,
+							// just continue with the next mirror...
+							continue
+						}
+
+						c.logger.Log("warning", "no existing mirror found for signaled git change", "mirror", mirrorName)
+						for _, hr := range hrs {
+							nsClient := c.client.HelmV1().HelmReleases(hr.Namespace)
+							_ = status.SetCondition(nsClient, *hr, status.NewCondition(
+								v1.HelmReleaseChartFetched,
+								corev1.ConditionUnknown,
+								ReasonGitNotReady,
+								"git mirror missing; starting mirroring again",
+							))
+							c.maybeMirror(mirrorName, hr.Spec.GitChartSource.GitURL)
+						}
+						// Wait for the signal from the newly requested mirror...
+						continue
+					}
+					if mirrorStatus, err := mirror.Status(); mirrorStatus != git.RepoReady {
+						c.logger.Log("warning", "mirror not ready for sync", "status", mirrorStatus, "mirror", mirrorName)
+						for _, hr := range hrs {
+							nsClient := c.client.HelmV1().HelmReleases(hr.Namespace)
+							_ = status.SetCondition(nsClient, *hr, status.NewCondition(
+								v1.HelmReleaseChartFetched,
+								corev1.ConditionUnknown,
+								ReasonGitNotReady,
+								err.Error(),
+							))
+						}
+						// Wait for the next signal...
+						continue
+					}
+					for _, hr := range hrs {
+						if _, ok := c.syncGitChartSource(mirror, hr); ok {
+							cacheKey, err := cache.MetaNamespaceKeyFunc(hr.GetObjectMeta())
+							if err != nil {
+								continue // this should never happen
+							}
+							c.releaseQueue.AddRateLimited(cacheKey)
+						}
+					}
+				}
+			case <-stopCh:
+				c.logger.Log("info", "stopping sync of git chart sources")
+				return
+			}
+		}
+	}()
+}
+
+// Load returns a pointer to the requested `GitChartSource` for the given
+// `v1.HelmRelease` if found, and a boolean indicating success.
+func (c *GitChartSourceSync) Load(hr *v1.HelmRelease) (*GitChartSource, bool) {
+
+	// Check if we have a source in store and return if it still equals
+	// to what is configured in the source.
+	c.sourcesMu.Lock()
```

Comment:
You may want to consider using a `RWMutex` and `RLock()`ing here instead of `Lock()`ing